File: src/main/java/org/galaxio/gatling/kafka/javaapi/protocol/KafkaProtocolBuilder.java
package org.galaxio.gatling.kafka.javaapi.protocol;

import io.gatling.core.protocol.Protocol;
import io.gatling.javaapi.core.ProtocolBuilder;
import org.galaxio.gatling.kafka.request.KafkaProtocolMessage;
import scala.Function1;

public class KafkaProtocolBuilder implements ProtocolBuilder {

    private org.galaxio.gatling.kafka.protocol.KafkaProtocolBuilder wrapped;

    public KafkaProtocolBuilder(org.galaxio.gatling.kafka.protocol.KafkaProtocolBuilder wrapped) {
        this.wrapped = wrapped;
    }

    public KafkaProtocolBuilder matchByValue() {
        this.wrapped = wrapped.matchByValue();
        return this;
    }

    public KafkaProtocolBuilder matchByMessage(Function1<KafkaProtocolMessage, byte[]> keyExtractor) {
        this.wrapped = wrapped.matchByMessage(keyExtractor);
        return this;
    }

    @Override
    public Protocol protocol() {
        return wrapped.build();
    }

}

File: src/main/scala/org/galaxio/gatling/kafka/protocol/KafkaProtocolBuilder.scala
package org.galaxio.gatling.kafka.protocol

import org.galaxio.gatling.kafka.protocol.KafkaProtocol.KafkaKeyMatcher

import scala.concurrent.duration.{DurationInt, FiniteDuration}

case object KafkaProtocolBuilder {

  def topic(name: String): KafkaProtocolBuilderPropertiesStep =
    KafkaProtocolBuilderPropertiesStep(name, Map.empty[String, Object])

  def requestReply: KafkaProtocolBuilderNew.type = KafkaProtocolBuilderNew

}

case class KafkaProtocolBuilder(topic: String, props: Map[String, Object], timeout: FiniteDuration = 60.seconds) {

  def timeout(t: FiniteDuration): KafkaProtocolBuilder = copy(timeout = t)

  def build = new KafkaProtocol(topic, props, props, timeout, KafkaKeyMatcher)

}
